R/MsExperiment-functions.R

```diff
@@ -404,3 +404,27 @@ readMsExperiment <- function(spectraFiles = character(),
     spectra(x) <- Spectra(spectraFiles, ...)
     linkSampleData(x, with = "sampleData.spectraOrigin = spectra.dataOrigin")
 }
+
+#' @title Consolidate links between samples and spectra after filtering
+#'
+#' @description
+#'
+#' If @spectra got filtered eventually present *links* between them and samples
+#' will no longer be valid and need to be updated/fixed. This function
+#' consolidates these links using a spectra variable `"._SPECTRA_IDX"` in
+#' `@spectra` that needs to represent/contain the index of the spectra
```

Comment:
So this `._SPECTRA_IDX` variable already existed before this update ?

Reply:
Ah, no. but very good question! I specifically add this variable **before** I call the filter function on the `Spectra` object. That's the only way (at least that I found) how I can now to which spectra the data was filtered. So, the workflow is:

- add the spectra variable `._SPECTRA_IDX` to the `Spectra` containing the index 1:length spectra (I chose this name for the spectra variable to not overwrite any potentially existing spectra variable with the same name - unlikely that somebody will call a spectra variable with this name - at least I assume)
- filter the `Spectra` object
- after filtering I need to update/consolidate the link (mapping) between samples and `Spectra` (which is stored in the `@sampleDataLinks[["spectra"]]` - it contains a two column matrix, the first with the index of the sample, the second with the index of the spectrum assigned to this sample). I can use the `._SPECTRA_IDX` variable for that, because this got also subset (along with the `Spectra` object). So, I need to only keep rows in the above mapping matrix with the spectrum index (second column) present in `._SPECTRA_IDX`.
- after that I remove the `._SPECTRA_IDX` spectra variable again, because it's not needed anymore.

let me know if something is still unclear.